core/src/main/java/io/grpc/internal/ClientCallImpl.java

```diff
@@ -165,19 +172,36 @@ public void start(final Listener<RespT> observer, Metadata headers) {
     checkNotNull(observer, "observer");
     checkNotNull(headers, "headers");
 
+    // Create the context
+    final Deadline effectiveDeadline = min(callOptions.getDeadline(), parentContext.getDeadline());
+    if (effectiveDeadline != parentContext.getDeadline()) {
+      context = parentContext.withDeadline(effectiveDeadline, deadlineCancellationExecutor);
+    } else {
+      context = parentContext.withCancellation();
+    }
+
+    // TODO: Context should cancel itself immediately when expired, so we don't need this check.
```

Comment:
Can you fix this in Context rather than here, no need to schedule the executor if its deadline already expired.